test/functional/server.py

```diff
@@ -2,29 +2,67 @@
 
 import argparse
 import http.server as server
+import os
 import ssl
 import sys
 import time
 
 
-class SlowResponse(server.BaseHTTPRequestHandler):
-    """Handler that returns data with a set delay between writes"""
+partial_download_file = ""
+slow_server = False
+
+class TestServer(server.BaseHTTPRequestHandler):
+    first_time = True
+
+    """Handler that returns data with a set delay between writes and/or
+       simulates a partial download on the first download attempt for the
+       file specified by partial_download_file."""
     def do_GET(self):
-        self.send_response(200)
+        try:
+            f = open(os.getcwd() + self.path, 'rb')
+        except:
+            self.send_response(404)
+            self.end_headers()
+            return
+
+        #Simulate partial download
+        partial_download = False
+        if self.path.endswith(partial_download_file) and TestServer.first_time:
+            partial_download = True
+            TestServer.first_time = False
+            self.send_response(206)
+        else:
+            self.send_response(200)
         self.send_header("Content-type", "text/html")
         self.end_headers()
-        response = "99990"
-        delay = 0.00001 # seconds
-        for c in response:
-            self.wfile.write(str.encode(c))
 
+        while True:
+            b = f.read(100)
```

Comment:
For the slow server, processing 100 bytes at a time is much different than 1 byte at a time. Perhaps:

```python
if slow_server:
    b = f.read(1)
else:
    b = f.read(100)
```

Reply:
This was written for the retry code and I used 100 because it was good enough to make suer versions and small manifests were downloaded all at one, but for sure testing with 1 will be better. Just make sure we're not forcing unnecessary retries. Not sure if 1 by 1 isn't too slow.

Reply:
Good point, I was primarily thinking about the check-update slow server test, which only asks for the `latest` file. Maybe the slow server could read 1 byte at a time for `latest`, and 100 at a time for everything else?

Reply:
So would passing the delay as an argument would work?

Reply:
More specifically, if the server writes 100 bytes for `latest`, that means that only one libcurl callback is needed to read the response, and the bug I linked above would not be triggered. By writing 1 byte at a time, likely more than one libcurl callback is needed. Also, the longer the version number in `latest`, the better; maybe the test should use a `latest` contains more than 5 bytes...

Reply:
I recorded the test execution times of update-slow-server when reading 1 byte and 100 bytes:

read(1): 134 seconds
read(100): 63 seconds

The test did not perform any unnecessary retries and it is normally skipped, so I will change this to f.read(1).

Reply:
nice. I'm good with that